src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs

```diff
@@ -0,0 +1,224 @@
+        /// allocated with the <see cref="System.Runtime.CompilerServices.RuntimeHelpers.AllocateTypeAssociatedMemory(Type, int)"/> API.
+        /// </remarks>
+        protected unsafe abstract ComInterfaceEntry* ComputeVtables(object obj, CreateComInterfaceFlags flags, out int count);
+
+        // Call to execute the abstract instance function
+        internal static unsafe void* CallComputeVtables(ComWrappers? comWrappersImpl, object obj, CreateComInterfaceFlags flags, out int count)
+            => (comWrappersImpl ?? s_globalInstance!).ComputeVtables(obj, flags, out count);
+
+        /// <summary>
+        /// Get the currently registered managed object or creates a new managed object and registers it.
+        /// </summary>
+        /// <param name="externalComObject">Object to import for usage into the .NET runtime.</param>
+        /// <param name="flags">Flags used to describe the external object.</param>
+        /// <returns>Returns a managed object associated with the supplied external COM object.</returns>
+        public object GetOrCreateObjectForComInstance(IntPtr externalComObject, CreateObjectFlags flags)
+        {
+            if (externalComObject == IntPtr.Zero)
+                throw new ArgumentNullException(nameof(externalComObject));
+
+            ComWrappers impl = this;
+            object? retValue = null;
+            GetOrCreateObjectForComInstanceInternal(ObjectHandleOnStack.Create(ref impl), externalComObject, flags, ObjectHandleOnStack.Create(ref retValue));
+
+            return retValue!;
+        }
+
+        [DllImport(RuntimeHelpers.QCall)]
+        private static extern void GetOrCreateObjectForComInstanceInternal(ObjectHandleOnStack comWrappersImpl, IntPtr externalComObject, CreateObjectFlags flags, ObjectHandleOnStack retValue);
+
+        /// <summary>
+        /// Create a managed object for the object pointed at by <paramref name="agileObjectRef"/> respecting the values of <paramref name="flags"/>.
+        /// </summary>
+        /// <param name="externalComObject">Object to import for usage into the .NET runtime.</param>
+        /// <param name="agileObjectRef">IAgileReference pointing at the object to import into the .NET runtime.</param>
+        /// <param name="flags">Flags used to describe the external object.</param>
+        /// <returns>Returns a managed object associated with the supplied external COM object.</returns>
+        /// <remarks>
+        /// The <paramref name="agileObjectRef"/> is an <see href="https://docs.microsoft.com/windows/win32/api/objidl/nn-objidl-iagilereference">IAgileReference</see> instance. This type should be used to ensure the associated external object, if not known to be free-threaded, is released from the correct COM apartment.
+        /// </remarks>
+        protected abstract object CreateObject(IntPtr externalComObject, IntPtr agileObjectRef, CreateObjectFlags flags);
```

Comment:
Should this take a desired `System.Type` of the wrapper as optional argument?

Reply:
- To save lookup time when the wrapper type is known
- (potentially) To disambiguate the cases where there are multiple candidates

Reply:
This suggestion aligns with something I had been considering with passing a "context". What are your thoughts about taking an optional `object` and the `ComWrappers` implementation could interpret that in an implementation dependent way.

Reply:
This suggestion aligns with something I had been considering with passing a "context". What are your thoughts about taking an optional object and the ComWrappers implementation could interpret that in an implementation dependent way.

Reply:
I got the idea for the `System.Type` argument by looking at existing System.Runtime.InteropServices.Marshal APIs, and validating whether a similar expressiveness can be achieved through this.

System.Object - I am not sure. If people need to smuggle extra arbitrary data, it can be done via ThreadStatic.

Reply:
Or is the scenario where the wrapper type is known covered by the RegisterObjectForInterface discussed in the other comment?

Reply:
I am not convinced of the need/use for the `RegisterObjectForInterface()`. I don't believe it has a use since it basically requires creation of an object that may be useless. In my opinion supplying the `Type` in this API makes much more sense.

Reply:
@jkotas What is the typical response for default parameters in new APIs? I would like to avoid users from having to provide `null`, but realize default parameters aren't always desirable. Also, see [here](https://github.com/dotnet/runtime/pull/32091#discussion_r382956773) for changing the `Type?` to `object?`.

Reply:
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/member-overloading says "DO use member overloading rather than defining members with default arguments". This is guideline from 2008. I believe that the current API design trend is more amicable to use of default arguments than this. I do not think there is a simple rule.